docs: CI pipeline documentation (#51)
feedzai/fairgbm: .github/ci_workflows_docs.md

# Ci scripts
The documentaion below showcases each CI workflows' functionality.

## .github/workflows/cuda.yml
This script aims to test the project's CUDA support with the following different configurations:
- Source build - Python 3.7, GCC compiler, CUDA 11.2.2
- Pip install - Python 3.8, Clang compiler, CUDA 10.0
- Wheel install - Python 3.9, GCC compiler, CUDA 9.0

It triggers on pushing and creating pull requests to master.
Set environment variables:
  - github_actions: 'true'
  - os_name: linux
  - task: cuda
  - conda_env: test-env 

> The steps are presented below

Reviewer: ```suggestion
The steps are presented below:
```

> ### Setup or update software on host machine

Reviewer: ```suggestion
### 1. Setup or update software on host machine
```

This step aims to setup the host machine with the necessary images and packages (Ubuntu image, Nvidia docker engine resources and docker)

> ### Setup and run tests

Reviewer: ```suggestion
### 2. Setup and run tests
```

This step aims to run the tests after the setup is done. This is achieved by setting up all the necessary variables and then running the docker image with any of the configuration combinations specified above.

## .github/workflows/linkchecker.yml

> As the name suggets, this step is meant to lint check. It is triggered with a cron schedule - once a day at 08:00am UTC. Otherwise, it can can be triggered manually.

Reviewer 2: ```suggestion
As the name suggests, this step is meant to link check. It is triggered with a cron schedule - once a day at 08:00am UTC. Otherwise, it can can be triggered manually.
```

Reviewer 3: Also `suggets` → `suggests` :)

Reviewer 2: Done :)

Set environment variables:
  - CONDA_ENV: test-env
  - GITHUB_ACTIONS: 'true'
  - OS_NAME: 'linux'
  - PYTHON_VERSION: 3.9
  - TASK: 'check-links'

> The steps are disclosed below.

Reviewer: ```suggestion
The steps are disclosed below:
```

> ### Checkout repository

Reviewer: ```suggestion
### 1. Checkout repository
```

> This step checks out the relevant repo's last 5 commits without any submodules

Reviewer 2: Can you give us any idea on why this happens? :)

> ### Setup and run tests

Reviewer: ```suggestion
### 2. Setup and run tests
```

Exports the relevant environment variables, and runs the [`.ci/setup.sh`](/.github/ci_scripts_docs.md#fairgbmcisetupsh) , which will install the necessary dependencies, and the [`.ci/test.sh`](/.github/ci_scripts_docs.md#fairgbmcitestsh), which will perform the link checking.

## .github/workflows/python_package.yml
This workflow is meant to test  the FairGBM package the following configuration combinations:
- sdist - Source distribution build (Python 3.9, Ubuntu)
- bdist - Binary distribution build (Python 3.8, Ubuntu)
- if-else - Tests conditional compilation paths (Python 3.8, Ubuntu)
- mpi + pip - MPI parallel learning via pip (Python 3.7, Ubuntu)
- mpi + wheel - MPI parallel learning via wheel (Python 3.7, Ubuntu)

This workflow is triggered on pushing and opening pull requests to the `master` and the `main-fairgbm`.

Set environment variables:
- CONDA_ENV: test-env
- GITHUB_ACTIONS: 'true'

> This workflow's steps are the following:

Reviewer: same as before, you get the idea :sweat_smile:

This workflow is meant to publish a python package into pypi. It is triggered on push with tags that match the `v*` expression.

The steps are explained below.

### Set up Python
This step sets up python 3.7.

### Install dependencies
As the name implies, this step install all the necesary dependencies.

### Build package
This step [builds](https://pypi.org/project/build/) the designated python package.

### Publish package
This step publishes the built package.

## .github/workflows/static_analysis.yml
This workflow performs static code analysis. It is triggered by pushing or creating pull requests to master.

The set environment variables are the following:
- COMPILER: 'gcc'
- CONDA_ENV: test-env
- GITHUB_ACTIONS: 'true'
- OS_NAME: 'linux'
- PYTHON_VERSION: 3.9

The steps are explained below. 

### Checkout repository
The first step in this task is to checkout the relevant action's repository.

### Setup and run tests
The following step runs the [`.ci/setup.sh`](/.github/ci_scripts_docs.md#fairgbmcisetupsh) and the [`.ci/test.sh`](/.github/ci_scripts_docs.md#fairgbmcitestsh) scripts. In this instance, this means to that the Python/C++ code is verified for style guidelines.

> ## r-check-docs

Reviewer 2: ```suggestion
### r-check-docs
```
